### PR TITLE
[AEP] docs: add proposal for `kwok` provider

### PR DESCRIPTION
**Pull request by vadasambar**
Signed-off-by: vadasambar <surajrbanakar@gmail.com>

#### What type of PR is this?

/kind documentation
/kind feature

#### What this PR does / why we need it:
This is a proposal for implementing `kwok` provider. Check the files for more explanation. 

#### Which issue(s) this PR fixes:

It will help fix issues like https://github.com/kubernetes/autoscaler/issues/5769

#### Does this PR introduce a user-facing change?

```release-note

```
Not sure what should be added ^ 
#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by vadasambar** [comment]
Note that the original plan was to create a first draft and then edit the proposal to match the [KEP template](https://github.com/kubernetes/enhancements/blob/master/keps/NNNN-kep-template/README.md) but after discussion in sig-autoscaling meeting yesterday, I found out that I don't need to write a proposal to add a cloud provider. 

I think having a proposal is still valuable (but it doesn't follow standard KEP template). It helps users understand the motivation for adding the cloud provider and it provides a document everyone can review and add comments/suggestions in. I will keep this PR open for reviews/suggestions from users/contributors before merging it in. 

I have asked `kwok` and `sig-scalability` for review:
* #kwok-dev: https://kubernetes.slack.com/archives/C04QPDWGW0M/p1687237015984919
* #sig-scalability: https://kubernetes.slack.com/archives/C09QZTRH7/p1687237346496719

File: cluster-autoscaler/proposals/kwok_cloud_provider.md [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+3. `kwok` provider as of now only understands GKE and EKS node labels. The plan is to expand this support for other cloud providers as we go. 
+### How does it look in action?
+You can check it [here](https://github.com/kubernetes/autoscaler/issues/5769#issuecomment-1590541506).
+
+### Definition of `DONE`
+* `kwok` provider is implemented
+* relevant unit tests have been added
+* `kwok` cloud provider is `demo`ed in one of the sig-autoscaling meetings
+
+### FAQ
```
[the remark this answers is not in the document]
Comment (vadasambar) [reply cluster-autoscaler/proposals/kwok_cloud_provider.md:145]: Thank you for the link. Fortunately, CA doesn't depend on `kubectl top pods` or `kubectl top nodes` metrics. It is based on kube state metrics (it uses pod CPU and memory requests metrics). 

That being said, having metrics server would definitely be nice. :smile:

File: cluster-autoscaler/proposals/kwok_cloud_provider.md [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+Implement in-tree cloud provider for [Kubernetes WithOut Kubelet](https://kwok.sigs.k8s.io/) (referred as `kwok` provider henceforth) by implementing the [`CloudProvider`](https://github.com/vadasambar/autoscaler/blob/0d167b4e0586b04e773079d81cb9c45bca9ce0d3/cluster-autoscaler/cloudprovider/cloud_provider.go#L98-L149) interface. 
+
+The `kwok` provider can be used by setting `--cloudprovider=kwok` in the CA flags. 
+
+### Function implementations
+* `BuildKwokCloudProvider`: 
+   * Create a lister to watch for nodes with the annotation `kwok.x-k8s.io/node: fake`
+      * The annotation is used to identify nodes created by the `kwok` provider (`kwok` provider creates nodes with this annotation). Limiting the lister `kwok` created nodes ensure we don't touch actual nodes. 
+   * Use `KWOK_TEMPLATES_PATH` env variable to fetch user defined template nodes. This is so that the users can specify their own custom template nodes. This can be the yaml result of `kubectl get nodes` from a different cluster with some changes. Or, it can be nodes defined by the user from scratch in yaml. The yaml has to be mounted in the CA pod at the path specified in `KWOK_TEMPLATES_PATH` env variable
```
[the remark this answers is not in the document]
Comment (vadasambar) [reply cluster-autoscaler/proposals/kwok_cloud_provider.md:98]: Thank you for sharing the link. https://github.com/kubernetes-sigs/kwok/issues/668 looks interesting. :+1: I

**Comment by vadasambar** [comment]
@wzshiming thank you for taking a look at my proposal :pray: :heart:

**Comment by x13n** [comment]
/assign

This looks quite interesting, I think some scalability issues could indeed be found with such cloud provider. Definitely not all of them, since some of them are cloud provider specific. I'd be happy to marge this, though maybe it would make sense to discuss on the sig meeting first? Or did that happen already? (I see this PR is quite old by now.)

**Comment by vadasambar** [comment]
> /assign
> 
> This looks quite interesting, I think some scalability issues could indeed be found with such cloud provider. Definitely not all of them, since some of them are cloud provider specific. I'd be happy to marge this, though maybe it would make sense to discuss on the sig meeting first? Or did that happen already? (I see this PR is quite old by now.)

Thank you for showing interest :heart: 

I already talked about this proposal in the sig meeting. Based on the discussion, I don't need to an AEP to add a new cloudprovider. I have kept this PR so that people can find it easily. 

I am working on getting the first release out. I plan to include this proposal as a doc in the PR. 

PR: https://github.com/kubernetes/autoscaler/pull/5820 

I think the code can improve (would love your review :) ). I want to get the first release out so that users can try it and work on the feedback.

**Comment by x13n** [comment]
Gotcha. I'd close this PR then since you're already working on the other one. Anyone reading this will know where to look next. Does that make sense?

Also, https://github.com/kubernetes/autoscaler/pull/5820 is currently marked as "DON'T REVIEW". Is it already in reviewable state?

**Comment by vadasambar** [comment]
> Gotcha. I'd close this PR then since you're already working on the other one. Anyone reading this will know where to look next. Does that make sense?

Agreed. I will close this PR. 
 
> Also, #5820 is currently marked as "DON'T REVIEW". Is it already in reviewable state?

I am writing tests right now. I would like to get a review on the PR in its current state but I will refrain (don't want to waste your time). Once I am done with the tests, I will mention you in the PR to let you know it is ready.

**Pull request closed without merging**